### PR TITLE
WIP: Add NBody code

### PR DESCRIPTION
**Pull request by miguelraz**
This needs a lot of work still - review needed.
I don't know if the NBodyProblem function loop code is to be written down in the `problems.jl` file, or if the lone constructor should do.

### CODE REVIEW

File: src/problems.jl [hunk 1]
```diff
@@ -1,13 +1,15 @@
+type AbstractPartitionedODEProblem  end
```
Comment (ChrisRackauckas) [inline src/problems.jl:1]: This already in DiffEqBase.jl...

File: src/problems.jl [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 # f = H(p,q)
 # Should create a conversion to a PartitionedODEProblem by autodifferentiation
 type HamiltonianProblem{uType,tType,isinplace,F,C} <: AbstractPartitionedODEProblem{uType,tType,isinplace}
-  f::F
-  u0::uType
-  tspan::Tuple{tType,tType}
-  callback::C
+  f :: F
+  u0:: uType
```
Comment (ChrisRackauckas) [inline src/problems.jl:6]: Don't mix grammar fixes with feature PRs

**Comment by ChrisRackauckas** [comment]
>I don't know if the NBodyProblem function loop code is to be written down in the problems.jl file, or if the lone constructor should do.

You should have a function that the problem uses that is built from the N that is passed in. Choose an implementation.

File: src/problems.jl (hunk 2 above)
Comment (miguelraz) [reply src/problems.jl:6]: Thanks, that seems reasonable.

**Comment by YingboMa** [comment]
Should `NBodyProblem` just a special case for `HamiltonianProblem`? See [this notebook](http://nbviewer.jupyter.org/github/JuliaDiffEq/DiffEqTutorials.jl/blob/master/PhysicalModels/Outer-Solar-System.ipynb). The assumption is that the potential energy doesn't depend on speed (not true for charged particle in magnetic field), and the mass matrix is diagonal. So that the Hamiltonian system can be simplified to

```
\ddot{q}=f(q)=−M \ ∇(V(q)).
```

Which is a `SecondOrderODEProblem`, and the user should just provide masses and potential field.

**Comment by ChrisRackauckas** [comment]
It should be a function which creates a `SecondOrderODEProblem`. Notice that `HamiltonianProblem` won't exist after https://github.com/JuliaDiffEq/DiffEqPhysics.jl/pull/13, since it's easier to just have them all create the respective problem types (`SecondOrderODEProblem`, `ODEProblem`).

http://aip.scitation.org/doi/pdf/10.1063/1.467523

That's a good paper which goes over a lot of details on symplectic integrators. They will never be able to handle the `V(t,q,p)` case, just `V(t,q)` case. However, Runge-Kutta Nystrom methods can. So there's no reason to disallow it, it just needs to be documented well with the solvers.

**Pull request closed without merging**